apps/guardian/src-tauri/src/main.rs

```diff
@@ -144,62 +440,675 @@ fn get_accounts(handle: AppHandle) -> Result<Vec<AccountAddress>, GetAccountsErr
+/// This has two potential outcomes:
+/// 1. All peer encrypted shares are valid, the secret share is computed and
+/// stored on disk and finally an OK signal is sent to the contract
+/// 2. One or more invalid keys have been detected, a complaint with the
+/// associated guardian accounts is registered in the election contract
+///
+/// The following steps are run:
+/// 1. Validate peer keys + generate encrypted shares if successful
+/// 2. Request approval of transaction proposal (either registration of shares
+/// or complaint)
+/// 3. Transaction submission + await finalization.
+///
+/// ## Errors
+/// - [`Error::NodeConnection`]
+/// - [`Error::NetworkError`]
+/// - [`Error::QueryFailed`]
+#[tauri::command]
+async fn generate_secret_share_flow<'a>(
+    channel_id: String,
+    active_guardian_state: State<'a, ActiveGuardianState>,
+    app_config_state: State<'a, AppConfigState>,
+    app_handle: AppHandle,
+    window: Window,
+) -> Result<(), Error> {
+    let cancel = handle_abort(&channel_id, &window);
+    let interaction = async {
+        let encrypted_shares =
+            match generate_secret_share(&active_guardian_state, &app_config_state, &app_handle)
+                .await
+            {
+                Ok(..) => Ok(()),
+                Err(Error::PeerValidation(accounts)) => Err(accounts),
+                Err(error) => return Err(error),
+            };
+
+        let mut contract = app_config_state.0.lock().await.connection().await?.contract;
+        let active_guardian_guard = active_guardian_state.0.lock().await;
+        let active_guardian = active_guardian_guard
+            .as_ref()
+            .ok_or(anyhow!("Active account not set"))?;
+
+        // Depending on whether any validation failures are detected, either:
+        // 1. register the generated shares
+        // 2. file a complaint with the guardian accounts with invalid key registrations
+        let guardian_status = match encrypted_shares {
+            Ok(..) => contract::GuardianStatus::VerificationSuccessful,
+            Err(accounts) => contract::GuardianStatus::SharesVerificationFailed(accounts),
+        };
+
+        let contract_update = contract
+            .dry_run_update::<contract::GuardianStatus, Error>(
+                "registerGuardianStatus",
+                Amount::zero(),
+                active_guardian.guardian.account,
+                &guardian_status,
+            )
+            .await?;
+        let ccd_cost = energy_to_ccd(contract_update.current_energy(), &app_config_state).await?;
+        let proposal = match guardian_status {
+            contract::GuardianStatus::VerificationSuccessful => {
+                ValidatedProposal::Success(ccd_cost)
+            }
+            contract::GuardianStatus::SharesVerificationFailed(_) => {
+                ValidatedProposal::Complaint(ccd_cost)
+            }
+            _ => unreachable!(), // As we know the guardian_status is one of the above
```

Comment:
This makes me sad, but I can see that it's hard to remove it without complications.

Reply:
I can return an `Error::Internal` instead, but my rationale here is, as before, that this is derived from a static value (or well.. you know what I mean) and that it would make sense to reserve `Err` cases to anything out of our control.

This is my preference, but if there is a reason why this is generally not a good approach, or if it is just idiomatically wrong, I'm open to changing this 🙂

Reply:
I guess I was not clear enough. I think this is fine, since this will really never happen.

It still makes me sad for aesthetic reasons.

Reply:
Haha 😭